### PR TITLE
[REEF-453]  Race condition in GroupCommunication Client initialization

### PR DESCRIPTION
**Pull request by jwang98052**
In GroupCommClient constructor, we register the task node to Name Server first, then instantiate CommunicationGroupClient, which will instantiate Operators and register handlers. In Operator constructors, we check if its parents and children are all registered based on the topology. There is a chance that a parent or a child was registered, but haven't completed the initialization, resulting a race condition.

This PR moves the registration in GroupCommClient as the last line in the constructor. And movs the initialize checking out of the Operator constructors. Initialize will be called when client gets Operator from CommunicationGroupClient. This would ensure all the registration happens after constructors. And registration checking happens after all the constructors.

JIRA: REEF-453(https://issues.apache.org/jira/browse/REEF-453)

This closes #

Author: Julia Wang  Email: jwang98052@yahoo.com

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs [hunk 1]
```diff
@@ -148,7 +158,9 @@ public IScatterSender<T> GetScatterSender<T>(string operatorName)
         /// <returns>The ScatterReceiver</returns>
         public IScatterReceiver<T> GetScatterReceiver<T>(string operatorName)
         {
-            return GetOperatorInstance<ScatterReceiver<T>>(operatorName);
+            var o = GetOperatorInstance<ScatterReceiver<T>>(operatorName);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs]: Couldn't the `WaitForRegistration()` call be done within `GetOperatorInstance`? That way, the change would be smaller.

File: lang/cs/Org.Apache.REEF.Network/NetworkService/WritableNetworkService.cs [hunk 2]
```diff
@@ -122,10 +127,10 @@ public void Register(IIdentifier id)
             _localIdentifier = id;
             NamingClient.Register(id.ToString(), _remoteManager.LocalEndpoint);
 
-            // Create and register incoming message handler
-            // TODO[REEF-419] This should use the TcpPortProvider mechanism
-            var anyEndpoint = new IPEndPoint(IPAddress.Any, 0);
-            _messageHandlerDisposable = _remoteManager.RegisterObserver(anyEndpoint, _messageHandler);
+            //// Create and register incoming message handler
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/NetworkService/WritableNetworkService.cs]: Please remove the commented out code.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IGroupCommOperator.cs [hunk 3]
```diff
@@ -39,5 +39,10 @@ public interface IGroupCommOperator<T>
         /// The operator version number.
         /// </summary>
         int Version { get; }
+
+        /// <summary>
+        /// Ensure all parent and children nodes in the topology are registered with teh Name Service.
+        /// </summary>
+        void WaitForRegistration();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IGroupCommOperator.cs]: Why can't this be the last thing an Operator does in its constructor?

**Comment by markusweimer** [comment]
@jwang98052 I left some questions.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IGroupCommOperator.cs (hunk 3 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IGroupCommOperator.cs]: Let me try to explain this. Before calling this function each task has to register itself with NameClient so that others can get out of their wait in the above function. Hence order of operation has to be: RegisterWithNetworkService followed by WaitForRegistration. Hence this cannot be called in operator constructor because then we have to call register before operator instantiation which will lead to Race Conditions. However, I think they can be called on the end of GroupCommClient constructor.

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IGroupCommOperator.cs]: Can't the flow in the constructors be:

  1. Register with name service.
  2. Call a `private` method `WaitForRegistration()`

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs]: I tried, without knowing the concrete T, it won't work.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IGroupCommOperator.cs (hunk 3 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IGroupCommOperator.cs]: The sequence must be
1. Complete all nested constructors to ensure handlers/observers are registered 
2. Then in GroupCommClient, Register with name service
3. Then call WaitForRegistration in each CommunicationGroupClient and each Operator. 

The problem was the operators is Dictionary<string, object>.  We Know the object is IGroupCommOperator of T, but we don't know T given the dictionary of operators. That was the reason I didn't call it in GroupCommClient constructor. 

However, I can use reflection to do it. The code change will follow.

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs]: I will use reflection and move the call to GroupCommClient constructor after Register to name client.

File: lang/cs/Org.Apache.REEF.Network/NetworkService/WritableNetworkService.cs (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/NetworkService/WritableNetworkService.cs]: done

**Comment by jwang98052** [comment]
Thanks for review. All review comments have been addressed and checked in. 
@dkm2110 : could you please build and rerun the tests in cluster with my change?

**Comment by dkm2110** [comment]
ok..I will pull you request and do it by tonight.

Dhruv

**Comment by dkm2110** [comment]
@markusweimer  @jwang98052  I looked at the pull request. Looks good to me. I also was able to build release version by doing some hacks suggested by Julia and the code works fine on cluster.

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs [hunk 4]
```diff
@@ -44,7 +44,7 @@ public class Timeout : Name<int>
         {
         }
 
-        [NamedParameter("Retry times", defaultValue: "5")]
+        [NamedParameter("Retry times", defaultValue: "500")]
         public class RetryCount : Name<int>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: Why this change? It seems orthogonal to the problem at hand. Also, this class can probably be `internal sealed` and would benefit from documentation.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs [hunk 5]
```diff
@@ -79,6 +76,17 @@ public sealed class BroadcastReceiver<T> : IBroadcastReceiver<T>
         public string OperatorName { get; private set; }
 
         /// <summary>
+        /// Ensure all parent and children nodes in the topology are registered with teh Name Service.
+        /// </summary>
+        public void WaitForRegistration()
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: Do these methods need to be `public`? They are purely internal to the group communications implementation, right?

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs [hunk 6]
```diff
@@ -93,7 +94,7 @@ public class CommunicationGroupClient : ICommunicationGroupClient
         /// <returns>The BroadcastSender</returns>
         public IBroadcastSender<T> GetBroadcastSender<T>(string operatorName)
         {
-            return GetOperatorInstance<BroadcastSender<T>>(operatorName);
+            return  GetOperatorInstance<BroadcastSender<T>>(operatorName);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs]: Pure formatting change. Please remove.

**Comment by markusweimer** [comment]
@jwang98052 I did another pass. One trivial change (a space), but also a question regarding the visibility of the `WaitForRegistration()` method introduced here.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs (hunk 5 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: It is defined in the interface, therefore must be public. Every sub class needs it. So must be defined in the interface. 

Another option is to add an abstract layer to put common stuff of the sub classes and remove it from the interface. There are a lot can be moved to a base layer. If that sounds a good way, I can do refactor in a separate Jira.

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs (hunk 6 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs]: ok

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs (hunk 4 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: This retry count is used in waiting for registration. When we have over 100 nodes, 5 times is far from enough. 
We don't want set it too be too low in case some nodes are just slow, if we simply throw an exception, that is not right. 
We don't want it to try endlessly in case some node is really dead, we should come out with exception. 
We want it to return as soon as all nodes in the group are registered, So increasing retry count is better then increasing sleep time. 
Dhruv and I discussed about it based on the cluster testing, this number would give 250 seconds, about 4 min waiting time before throwing exception.

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: If we don't want clients to set it, yes, it can be internal sealed.

**Comment by jwang98052** [comment]
addressed review comments and pushed in.

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs (hunk 4 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: Just setting it to an arbitrary number that happens to work seems dangerous. We should at least add the above reasoning to the comments of the code to make sure we remember why we choose 500 as the magic number.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs (hunk 5 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: We could have an `internal` interface that extends the `public` one and only adds this one `internal` method, right? That way, we don't have to advertise this method into the IntelliSense of our users.

**Comment by markusweimer** [comment]
@jwang98052 Thanks, I left some more comments :)

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs (hunk 4 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: Actually speaking this number should be some monotonic function of number
of nodes in the communication group. More the number of nodes, greater time
it takes for driver to get all evaluators running, greater time it takes to
set up connection due to latencies and greater the chance of hitting a slow
node.

Dhruv

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: All the more reason to resolve this now. Maybe this number needs to be provided by an injectable interface? That way, it could be computed at runtime.

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: The number is a named parameter that can be configured. I have changed it to be internal sealed. And added comments.

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: If we would like more fancy algorithm, we can do it in a separate jira.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs (hunk 5 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: Yes, I can add an internal interface.

**Comment by jwang98052** [comment]
Review  comments addressed and pushed in.

**Comment by markusweimer** [comment]
I'll do another pass.

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs [hunk 7]
```diff
@@ -44,18 +44,25 @@ public class Timeout : Name<int>
...
-
-        [NamedParameter("sleep time to wait for handlers to be registered", defaultValue: "500")]
-        public class SleepTimeWaitingForHandler : Name<int>
-        {
-        }
-
-        [NamedParameter("Retry times to wait for handlers to be registered", defaultValue: "5")]
-        public class RetryCountWaitingForHanler : Name<int>
+        /// <summary>
+        /// Each COmmunication group needs to check and wait untill all the other nodes in the group are registered to the NameServer
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: Typo.

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs [hunk 8]
```diff
@@ -44,18 +44,25 @@ public class Timeout : Name<int>
...
+        /// Sleep time is set between each retry. 
+        /// </summary>
+        [NamedParameter("sleep time to wait for nodes to be registered", defaultValue: "500")]
+        internal sealed class SleepTimeWaitingForRegistration : Name<int>
+        {
+        }
+
+        /// <summary>
+        /// Each COmmunication group needs to check and wait untill all the other nodes in the group are registered to the NameServer
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: Typo. Also, much of this could go into the `remarks`.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs [hunk 9]
```diff
@@ -116,5 +113,16 @@ public T Receive()
 
             return PipelineDataConverter.FullMessage(messageList);
         }
+
+        /// <summary>
+        /// Ensure all parent and children nodes in the topology are registered with teh Name Service.
+        /// </summary>
+        void IGroupCommOperatorInternal.WaitForRegistration()
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: is this implicitly `internal` because the interface is?

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ScatterReceiver.cs [hunk 10]
```diff
@@ -99,5 +96,16 @@ public List<T> Receive()
             _topology.ScatterToChildren(elements, MessageType.Data);
             return elements.ToList();
         }
+
+        /// <summary>
+        /// Ensure all parent and children nodes in the topology are registered with teh Name Service.
+        /// </summary>
+        void IGroupCommOperatorInternal.WaitForRegistration()
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ScatterReceiver.cs]: The other implementations don't prefix it with `IGroupCommOperatorInternal.`. Is this needed here?

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs [hunk 11]
```diff
@@ -175,5 +177,18 @@ public IScatterReceiver<T> GetScatterReceiver<T>(string operatorName)
...
+        /// <summary>
+        /// Call each Operator to easure all the nodes in the topology group has been registered
+        /// </summary>
+        void ICommunicationGroupClientInternal.WaitingForRegistration()
+        {
+            foreach (var op in _operators.Values)
+            {
+                var methods = op.GetType().GetMethods(BindingFlags.NonPublic | BindingFlags.Instance);
+                var queriedMethods = methods.Where(e => e.Name.Equals("Org.Apache.REEF.Network.Group.Operators.IGroupCommOperatorInternal.WaitForRegistration")).ToArray();
+                queriedMethods[0].Invoke(op, null);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs]: Is this neccesary? We might make `IGroupCommOperatorInternal` extend the `IGroupCommOperator` interface, right? And in that case, we can call the method directly? Or am I missing something?

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/GroupCommClient.cs [hunk 12]
```diff
@@ -69,6 +68,13 @@ public sealed class GroupCommClient : IGroupCommClient
                 ICommunicationGroupClient commGroupClient = groupInjector.GetInstance<ICommunicationGroupClient>();
                 _commGroups[commGroupClient.GroupName] = commGroupClient;
             }
+
+            networkService.Register(new StringIdentifier(taskId));
+
+            foreach (var group in _commGroups.Values)
+            {
+                ((ICommunicationGroupClientInternal)group).WaitingForRegistration();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/GroupCommClient.cs]: Can't we change the type of `group` to be `ICommunicationGroupClientInternal`?

**Comment by markusweimer** [comment]
@jwang98052 I've done another pass. Minor typos and some C# questions for my education :-)

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs (hunk 9 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: Yes. The interface is internal, user cannot access it from outside.

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs (hunk 8 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: Updated

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs (hunk 7 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: done

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ScatterReceiver.cs (hunk 10 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ScatterReceiver.cs]: All the Operators that implement this internal interface must prefix it, otherwise, get build error.

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs (hunk 11 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/CommunicationGroupClient.cs]: IGroupCommOperatorInternal cannot extend IGroupCommOperator because we have one more layer like IBroadcastReceiver, IBroadcastSender which are public. They cannot extend from internal interface. 
The reason I am using reflection is because we have no knowledge of T for the Operators in operator dictionary. They are objects. Even if we have a way to use a proper interface, because we don't know T, we are still not able to call the method.

File: lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/GroupCommClient.cs (hunk 12 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/GroupCommClient.cs]: Yes. Need to cast after GetInstance.

**Comment by jwang98052** [comment]
review comments addressed and pushed in.

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs [hunk 13]
```diff
@@ -44,18 +44,27 @@ public class Timeout : Name<int>
...
-
-        [NamedParameter("sleep time to wait for handlers to be registered", defaultValue: "500")]
-        public class SleepTimeWaitingForHandler : Name<int>
-        {
-        }
-
-        [NamedParameter("Retry times to wait for handlers to be registered", defaultValue: "5")]
-        public class RetryCountWaitingForHanler : Name<int>
+        /// <summary>
+        /// Each COmmunication group needs to check and wait until all the other nodes in the group are registered to the NameServer
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: There is still a typo here. `COmmunication` --> `Communication`

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs (hunk 9 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: OK, thanks!

**Comment by markusweimer** [comment]
@jwang98052 Did another pass, two minor typos.

File: lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs [hunk 14]
```diff
@@ -44,18 +44,27 @@ public class Timeout : Name<int>
...
+        /// Sleep time is set between each retry. 
+        /// </summary>
+        [NamedParameter("sleep time to wait for nodes to be registered", defaultValue: "500")]
+        internal sealed class SleepTimeWaitingForRegistration : Name<int>
+        {
+        }
+
+        /// <summary>
+        /// Each COmmunication group needs to check and wait until all the other nodes in the group are registered to the NameServer
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Config/GroupCommConfigurationOptions.cs]: Same typos as above.

**Comment by jwang98052** [comment]
Updated and rebased and force pushed into REEF-453. Please pull to a fresh local branch when merging.

**Comment by markusweimer** [comment]
This looks good, I'll test

**Comment by jwang98052** [comment]
After rebase, the build is broken as the issue described in REEF-466. You have to merge REEF-466 first, then merge to REEF-453 and testing. 
Debug build in master is broken now until REEF-466 is in.

**Comment by markusweimer** [comment]
@jwang98052 Yes, I noticed. REEF-466 is in now and I'll test this.

**Comment by markusweimer** [comment]
tests pass, I will merge
